text/0615-autotracking-memoization.md

```diff
@@ -0,0 +1,385 @@
+autotracked state.
+
+Data layers like Ember Data could also benefit from this capability. These
+layers tend to have to keep state in sync between multiple levels of caching,
+which traditionally was done with computed properties and eventing systems. The
+ability to use autotracking to replace these systems, and to define their own
+reactive semantics, could help complex libraries and data layers out immensely.
+
+## Detailed design
+
+This RFC proposes four functions to be added to Ember's public API:
+
+```ts
+interface Cache<T = unknown> {}
+
+function createCache<T>(fn: () => T): Cache<T>;
+
+function getValue<T>(cache: Cache<T>): T
+
+function isCache(maybeCache: object): boolean;
+
+function isConst(cache: Cache): boolean;
+```
+
+These functions are exposed as exports of the `@glimmer/tracking/primitives/cache`
+module:
+
+```ts
+import {
+  createCache,
+  getValue,
+  isCache,
+  isConst,
+} from '@glimmer/tracking/primitives/cache';
+```
+
+### Usage
+
+`cache` receives a function, and returns a cache instance for that function.
+Users can call `getValue()` on the cache instance to run the function and get
+the value of its output. The cache will then return the same value whenever
+`getValue` is called again, until one of the tracked values that was _consumed_
+while it was running previously has been _dirtied_.
+
+```ts
+class State {
+  @tracked value;
+}
+
+let state = new State();
+let count = 0;
+
+let counter = createCache(() => {
+  // consume the state
+  state.value;
+
+  return ++count;
+});
+
+getValue(counter); // 1
+getValue(counter); // 1
+
+state.value = 'foo';
+
+getValue(counter); // 2
+```
+
+Getting the value of a cache also _consumes_ the cache. This means caches can be
+nested, and whenever you use a cache inside of another cache, the outer cache
+will dirty if the inner cache dirties.
+
+```ts
+let inner = createCache(() => { /* ... */ })
+
+let outer = createCache(() => {
+  /* ... */
+
+  inner();
+});
+```
+
+This can be used to break up different parts of a execution so that only the
+pieces that changed are rerun.
+
+The `isCache` function can be used to determine if a value is a cache, in cases
+where it's not possible to know if the value is a cache or not.
+
+```js
+let cache = createCache(() => {});
+let notACache = () => {};
+
+isCache(cache); // true
+isCache(notACache); // false
+```
+
+### Constant Caches
+
+Caches will only recompute if any of the tracked inputs that were consumed
+previously change. If there _were_ no consumed tracked inputs, then they will
+never recompute.
+
+```ts
+let count = 0;
+
+let counter = createCache(() => {
+  return ++count;
+});
+
+getValue(counter); // 1
+getValue(counter); // 1
+getValue(counter); // 1
+
+// ...
+```
+
+When this happens, it often means that optimizations can be made in the code
+surrounding the computation. For instance, in the Glimmer VM, we don't emit
+updating bytecodes if we detect that a memoized function is constant, because
```

Comment:
It may make sense to further define `is constant`.  I think it shares the same properties in meaning as the JS construct "constant", but could be also gain further meaning.  Does it also mean "steady" or "invariable"?

Reply:
Perhaps an additional adjective could be used in the docs section if you felt there was an improvement!

Reply:
```suggestion
updating bytecodes if we detect that a memoized function can never change, because
```